### ISSUE TITLE
Aliased functions cannot be properly debugged, because of the use of eval

### ISSUE THREAD

**Issue by janmartenjongerius** [DivineOmega/php-camel-caser#1]
[As user phordijk on Reddit explains](https://www.reddit.com/r/PHP/comments/9hjsiu/php_camel_caser_lets_you_use_builtin_php/e6cm2jj):

> Even when implementing this functions eval()ing code like that to create the function will totally fuck up your error messages. Good luck debugging your code.

Language aside, this person has a great point about being able to read back the correct information when debugging code.

To address this, [I wrote a proof of concept that loads the aliases from file](https://3v4l.org/jtdJE):

```php
$code = array_reduce(
    get_defined_functions()['internal'] ?? [],
    function (string $carry, string $original): string {
        $alias = lcfirst(
            str_replace(
                ' ',
                '',
                ucwords(
                    str_replace(
                        ['-', '_'],
                        ' ',
                        $original
                    )
                )
            )
        );
        
        if ($alias != $original
            && !empty($alias)
            && !function_exists($alias)
        ) {
            // @todo Use reflection on the original to inherit
            // parameter list.
            $carry .= sprintf(
                <<<'FUNCTION'
if (!function_exists('%1$s')) {
    /**
     * @see %2$s
     */
    function %1$s (...$arguments) {
        return %2$s(...$arguments);
    }
}


FUNCTION
,
                $alias,
                $original
            );
        }
        
        return $carry;
    },
    '<?php' . PHP_EOL
);

// This should be solved with some sort of caching mechanism, to prevent
// having to do this every time.
$storage = tempnam(sys_get_temp_dir(), 'camelCasedInternalFunctions');

file_put_contents($storage, $code);
require $storage;
```
The code above solves the following:

* No dependency on `illuminate/support`
* No use of the `eval` function
  * No runtime evaluation of code
  * Ability to identify aliased functions in stack traces

About the cache:  one could invalidate it by making a hash out of the available internal functions, like `$hash = md5(implode(':', get_defined_functions()['internal'] ?? []));` and then to incorporate this in the file name of the cache file. That way, if a user installs or removes a PHP extension or library, the cache is invalidated automatically.

```php
<?php
$storage = sprintf(
    __DIR__ . '/aliases.%s.php',
    md5(
        implode(
            ':',
            get_defined_functions()['internal'] ?? []
        )
    )
);
```

If you agree with the points made in this issue and would like help implementing this, I'm happy to help.

**Comment by Jord-JD** [comment]
I definitely like the ideas behind this. :+1: I'm not sure about storing the resulting file in the system temporary directory, as it could potentially be edited by others resulting in a local code execution vulnerability. This would need thinking about.

Other than that, really great ideas here. Definitely feel free to open a PR. :smile:

**Comment by janmartenjongerius** [comment]
@DivineOmega so, how about if I would implement it using an approach where it is stored in the local vendor directory, with a hash that corresponds to the available internal functions, as described in the lower part of this issue? To make this testable code, I suggest to split it up, so one part of the code will be responsible for creating aliases of a given list of function names, one part will be responsible for determining the aliases and one part will be responsible for caching and loading the functions. That would allow unit tests that don't depend on the locally installed functions and an integration test to see if all possible candidates are available in aliased form after the code is initialized. Users could then reuse this code to quickly create aliases for their own functions as well.

If you see merit in these approaches, I will work on these modifications. Can you suggest a namespace you would prefer this code to live in?

**Comment by Jord-JD** [comment]
The vendor directory is probably the best idea, though I have occasionally experienced permissions issue writing to `vendor/` on some platforms.

Definitely agree with splitting things up to enable testing.

Perhaps `DivineOmega\CamelCaser\` for the namespace.

**Comment by janmartenjongerius** [comment]
While developing this feature locally, I encountered [a bug in PHP which causes the function `\hash_update_file` to be incompatible](https://bugs.php.net/bug.php?id=76918).

**Comment by janmartenjongerius** [comment]
I've pushed the required changes in [feature/1](https://github.com/johmanx10/php-camel-caser/tree/feature/1). It works as intended, but I haven't written tests and the `README.md` could do with an update to explain the separate components that are now available. Lastly, I intend to add some example files and a `.gitattributes` file to ensure test and example files, as well as configuration files, are kept from dist-archives on Github.

@DivineOmega could you verify that the feature branch is working for you as expected?

**Comment by Jord-JD** [comment]
@johmanx10 I've just quickly reviewed the code, and so far looks great! Thank you. I shall give it a proper test later today. Feel free to open a WIP PR from your branch.

**Comment by Jord-JD** [comment]
@johmanx10 Just tested. This all works as expected. Thanks! :slightly_smiling_face:

**Comment by Jord-JD** [comment]
@johmanx10 I've just merged #2, which offers similar functionality to what you've been working on. However, I'm still very interested in your solution along with the related unit tests if you get the chance to work on it any further.
